### PR TITLE
POC Merge concurrent Lookup requests.

### PR DESCRIPTION
**Pull request by mykhailo-kuchma**
Relates-To: OLPEDGE-1805

Signed-off-by: Mykhailo Kuchma <ext-mykhailo.kuchma@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp [hunk 1]
```diff
@@ -51,13 +52,35 @@ std::string GetDatastoreServerUrl(const std::string& partition) {
...
+  }
+
+ private:
+  std::mutex mutex_;
+  std::unordered_map<std::string, std::mutex> mutexes_;
+};
+
+// A temporary solution to avoid multiple online requests to lookup service for
+// the same catalog.
+static LookupApiLock gLookupLock;
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp:72]: I am a little puzzled by the coding style here. If I read this correct here https://google.github.io/styleguide/cppguide.html#Constant_Names then also static variables should be names like constants, means `kLookupLock`. Right?

File: olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp [hunk 2]
```diff
@@ -51,13 +52,35 @@ std::string GetDatastoreServerUrl(const std::string& partition) {
...
     const client::HRN& catalog,
     client::CancellationContext cancellation_context, std::string service,
     std::string service_version, FetchOptions options,
     client::OlpClientSettings settings) {
+  // This lock is required to avoid concurrent requests to online.
+  std::unique_lock<std::mutex> lock = gLookupLock.Lock(catalog);
+
   repository::ApiCacheRepository repository(catalog, settings.cache);
 
   if (options != OnlineOnly && options != CacheWithUpdate) {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp:86]: The mutex is never deleted after the response comes? Is this intentional?

File: olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp [hunk 3]
```diff
@@ -51,13 +52,35 @@ std::string GetDatastoreServerUrl(const std::string& partition) {
              ? "https://api-lookup." + server_url->second + "/lookup/v1"
              : "";
 }
+
+class LookupApiLock {
+ public:
+  std::unique_lock<std::mutex> Lock(const client::HRN& catalog) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    const std::string& key = catalog.ToString();
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp:60]:
```suggestion
    const auto key = catalog.ToString();
```

File: olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp [hunk 4]
```diff
@@ -51,13 +52,35 @@ std::string GetDatastoreServerUrl(const std::string& partition) {
...
+    std::unique_lock<std::mutex> catalog_lock(mutexes_[key]);
+    return catalog_lock;
+  }
+
+ private:
+  std::mutex mutex_;
+  std::unordered_map<std::string, std::mutex> mutexes_;
+};
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp:68]: I would rather create some helper class that is wrapped around a static mutex/unordered_map and automatically locks or waits for lock internally and removes the mutex from the unordered_map when it leaves the scope.

File: olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp [hunk 5]
```diff
@@ -58,6 +60,16 @@ ApiClientLookup::ApiClientResponse ApiClientLookup::LookupApi(
     client::CancellationContext cancellation_context, std::string service,
     std::string service_version, FetchOptions options,
     client::OlpClientSettings settings) {
+
+  // This lock is required to avoid concurrent requests to online.
+  repository::NamedMutex mutex(catalog.ToString());
+  std::lock_guard<repository::NamedMutex> lock(mutex);
+
+  // If we are not planning to go online, release the lock.
+  if (options == CacheOnly) {
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-read/src/ApiClientLookup.cpp:69]: If CacheOnly and in case we have no cache.

**Comment by mykhailo-kuchma** [comment]
@andescu Created a proper ticket, and another branch - https://github.com/heremaps/here-data-sdk-cpp/pull/858

**Pull request closed without merging**